app/controllers/concerns/application_shared.rb

```diff
@@ -21,4 +21,77 @@ def set_timezone
   def current_permission
     [action_permission, controller_permission].join('_')
   end
+
+  def set_taxonomy
+    user = User.current
+    return if user.nil?
+
+    ['location', 'organization'].each do |taxonomy|
+      next unless Taxonomy.enabled?(taxonomy.to_sym)
+
+      available = user.send("my_#{taxonomy.pluralize}")
+      determined_taxonomy = nil
+
+      # this keeps "any context" for admins unless they ask for specific context via parameters
+      if api_request? && user.admin?
+        if params["#{taxonomy}_id"].present?
+          determined_taxonomy = available.where(:id => params["#{taxonomy}_id"]).first
+
+          # in case user asked for taxonomy that does not exist or is not accessible, we reply with 404
+          if determined_taxonomy.nil?
+            not_found _("%{taxonomy} with id %{id} not found") % { :taxonomy => taxonomy.capitalize, :id => params["#{taxonomy}_id"] }
+            return false
+          end
+        else
+          determined_taxonomy = nil
+        end
+      # if there was a session set, try to load the taxonomy from there
+      elsif request.session["#{taxonomy}_id"].present?
+        determined_taxonomy = available.where(:id => request.session["#{taxonomy}_id"]).first
+        # warn user if taxonomy stored in session does not exist and delete it from session (probably taxonomy has been deleted meanwhile)
+        if determined_taxonomy.nil?
+          warning _("#{taxonomy.capitalize} you had selected as your context has been deleted") unless api_request?
```

Comment:
There are already 3 conditions that check for `api_request` here. And I would argue there should be 4, as via API, I would expect for this case also to receive `not_found` instead of having a default selected. Please reconsider having two methors: for UI and for API, if nothing else, it would reduce this checks on the `api_request`, which make it quite hard to understand.